async/kafka/kafka_test.go

```diff
@@ -315,3 +317,261 @@ func TestConsumer_ConsumeWithoutGroup(t *testing.T) {
+				versionedConsumerMessage("\"key\" \"value\"]", &sarama.RecordHeader{}, 1),
+				// will use json decoder based on the message header
+				versionedConsumerMessage("[\"key\",\"value\"]", &sarama.RecordHeader{}, 1),
+				// will fail at the result level due to the wrong message header, string instead of json
+				versionedConsumerMessage("[\"key\",\"value\"]", &sarama.RecordHeader{}, 2),
+				// will use void decoder because there is no message header
+				versionedConsumerMessage("any string ... ", &sarama.RecordHeader{}, 99),
+				// will produce error due to the content type invoking the erroringDecoder
+				versionedConsumerMessage("[\"key\",\"value\"]", &sarama.RecordHeader{}, 9),
+				// will use string decoder based on the message header
+				versionedConsumerMessage("key value", &sarama.RecordHeader{}, 2),
+			},
+			dmsgs:   [][]string{{"key", "value"}, {}, {"key", "value"}},
+			decoder: combinedDecoder,
+		},
+	}
+
+	for _, testdata := range testingdata {
+		testMessageClaim(t, testdata)
+	}
+
+}
+
+func saramaConsumerMessage(value string, header *sarama.RecordHeader) *sarama.ConsumerMessage {
+	return versionedConsumerMessage(value, header, 0)
+}
+
+func versionedConsumerMessage(value string, header *sarama.RecordHeader, version uint8) *sarama.ConsumerMessage {
+
+	bytes := []byte(value)
+
+	if version > 0 {
+		bytes = append([]byte{version}, bytes...)
+	}
+
+	return &sarama.ConsumerMessage{
+		Topic:          "TEST_TOPIC",
+		Partition:      0,
+		Key:            []byte("key"),
+		Value:          bytes,
+		Offset:         0,
+		Timestamp:      time.Now(),
+		BlockTimestamp: time.Now(),
+		Headers:        []*sarama.RecordHeader{header},
+	}
+}
+
+func testMessageClaim(t *testing.T, data decodingTestData) {
+
+	ctx := context.Background()
+
+	counter := eventCounter{}
+
+	factory, err := New("name", "topic", "group", []string{"0.0.0.0:9092"}, data.decoder)
+
+	assert.NoError(t, err, "Could not create factory")
+
+	c, err := factory.Create()
+
+	assert.NoError(t, err, "Could not create component")
+
+	// do a dirty cast for the sake of facilitating the test
+	kc := reflect.ValueOf(c).Elem().Interface().(consumer)
+
+	// claim and process the messages and update the counters accordingly
+	for _, km := range data.msgs {
+
+		if data.combinedDecoderVersion != 0 {
+			km.Value = append([]byte{byte(data.combinedDecoderVersion)}, km.Value...)
+
+		}
+
+		msg, err := claimMessage(ctx, &kc, km)
+
+		if err != nil {
+			counter.claimErr++
+			continue
+		}
+
+		err = process(&counter, &data)(msg)
+		if err != nil {
+			println(fmt.Sprintf("Could not process message %v : %v", msg, err))
+		}
+	}
+
+	assert.Equal(t, data.counter, counter)
+
+}
+
+// some naive decoder implementations for testing
+
+func erroringDecoder(data []byte, v interface{}) error {
+	return fmt.Errorf("Predefined Decoder Error for message %s", string(data))
+}
+
+func VoidDecoder(data []byte, v interface{}) error {
```

Comment:
If it is not exported then ...
```suggestion
func voidDecoder(data []byte, v interface{}) error {
```